File: app/dao/knowledgeIndex.ts
import fs from 'fs/promises';
import { fileURLToPath } from 'node:url';
import path from 'path';
import dotenv from 'dotenv';
import {
  VectorStoreIndex,
  OpenAI,
  OpenAIEmbedding,
  IngestionPipeline,
  storageContextFromDefaults,
  MetadataFilters,
  Document,
  BaseNode,
  Metadata,
  TitleExtractor,
} from 'llamaindex';
import { PineconeVectorStore } from '@llamaindex/pinecone';
import { SimpleDirectoryReader } from '@llamaindex/readers/directory';
import { Index, Pinecone, RecordMetadata } from '@pinecone-database/pinecone';

import { Settings } from '@llamaindex/core/global';
import { loadDocuments } from '@/app/dao/utils.server';
import { SentenceSplitter } from '@llamaindex/core/node-parser';
import { EngineResponse } from '@llamaindex/core/schema';
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
// Load environment variables from a specific .env file
dotenv.config({ path: path.resolve(__dirname, '../../.env') });
Settings.llm = new OpenAI();
Settings.embedModel = new OpenAIEmbedding({
  model: 'text-embedding-3-small',
  dimensions: 1536,
});

async function getSourceFilenames(sourceDir: string) {
  return await fs
    .readdir(sourceDir)
    .then((fileNames) => fileNames.map((file) => sourceDir + '/' + file));
}

function callback(
  category: string,
  name: string,
  status: unknown,
  message: string = ''
): boolean {
  console.log(category, name, status, message);
  return true;
}

class KnowledgeIndex {
  private client: Pinecone;
  private pc_index: Index<RecordMetadata>;
  public constructor(indexName: string) {
    this.client = new Pinecone({
      apiKey: process.env.PINECONE_API_KEY || 'your-api-key',
    });
    this.pc_index = this.client.Index(indexName);
  }

  public async loadFromDirectories(
    absoluteSourceDir: string
  ): Promise<VectorStoreIndex | undefined> {
    // const sourceDir = '/home/jingyi/WebstormProjects/copilot-v2/public/uploads';
    const fileList = await getSourceFilenames(absoluteSourceDir);
    const count = fileList.length;
    console.log(`Found ${count} files`);

    try {
      const rdr = new SimpleDirectoryReader(callback);
      const docs = await rdr.loadData({ directoryPath: absoluteSourceDir });
      const pcvs = new PineconeVectorStore({
        indexName: 'sources',
        namespace: 'lama',
        chunkSize: 512,
      });

      const ctx = await storageContextFromDefaults({ vectorStore: pcvs });

      return await VectorStoreIndex.fromDocuments(docs, {
        storageContext: ctx,
      });
    } catch (err) {
      console.error('Error in load directories', err);
      return undefined;
    }
  }

  //TODO: what are different indexing strategies from llamaIndex
  public async uploadToIndex(
    sourceMeta: {
      sourceId: string;
      sourceName: string;
    },
    buffer: Buffer<ArrayBufferLike>
  ): Promise<VectorStoreIndex | undefined> {
    const { sourceId, sourceName } = sourceMeta;

    try {
      const document = new Document({
        text: buffer.toString('utf-8'),
        metadata: {
          source_id: sourceId,
          source_name: sourceName,
        },
      });

      const pcvs = new PineconeVectorStore({
        indexName: 'sources',
        namespace: 'buffers',
        chunkSize: 1024,
      });

      const ctx = await storageContextFromDefaults({ vectorStore: pcvs });

      const resp = await VectorStoreIndex.fromDocuments([document], {
        storageContext: ctx,
      });

      return resp;
    } catch (err) {
      console.error(err);
      throw err;
    }
  }

  public async query(query: string) {
    const pcvs = new PineconeVectorStore({
      indexName: 'sources',
      namespace: 'buffers',
      chunkSize: 512,
    });

    const index = await VectorStoreIndex.fromVectorStore(pcvs);

    return index.asQueryEngine().query({ query });
  }

  async queryDocuments(
    documentIds: string[],
    query: string
  ): Promise<EngineResponse> {
    const pcvs = new PineconeVectorStore({
      indexName: 'sources',
      namespace: 'buffers',
      chunkSize: 512,
    });

    const index = await VectorStoreIndex.fromVectorStore(pcvs);

    const filters: MetadataFilters = {
      filters: [
        {
          key: 'source_id',
          value: documentIds,
          operator: 'in',
        },
      ],
    };

    const queryEngine = index.asQueryEngine({
      preFilters: filters,
      similarityTopK: 5,
    });

    return queryEngine.query({ query });
  }

  async summarizeSource(source_id: string) {
    const pcvs = new PineconeVectorStore({
      indexName: 'sources',
      namespace: 'buffers',
      chunkSize: 512,
    });

    const index = await VectorStoreIndex.fromVectorStore(pcvs);

    const filters: MetadataFilters = {
      filters: [
        {
          key: 'source_id',
          value: [source_id],
          operator: 'in',
        },
      ],
    };

    const queryEngine = index.asQueryEngine({
      preFilters: filters,
      similarityTopK: 10,
      // retriever: index.asRetriever({ mode: SummaryRetrieverMode.LLM }),
    });

    return queryEngine.query({ query: 'summarize the article' });
  }

  async hardDeleteSource(ref_doc_id: string) {
    try {
      const pcvs = new PineconeVectorStore({
        indexName: 'sources',
        namespace: 'buffers',
        chunkSize: 512,
      });

      const index = await VectorStoreIndex.fromVectorStore(pcvs);

      console.log('index', index);

      const resp = await index.deleteRefDoc(ref_doc_id);

      console.log(resp);
    } catch (error) {
      if (error instanceof Error) {
        console.error(
          `Error deleting doc_id ${ref_doc_id} from vector db `,
          error.message
        );
      } else {
        console.error('Unexpected error', error);
      }
      throw error;
    }
  }

  async listAllChunksForDocument(source_id: string) {
    try {
      const ns = this.pc_index.namespace('buffers');
      const res = await ns.list({
        source_id: { $eq: source_id },
      });
      console.log(res);
    } catch (error) {
      if (error instanceof Error) {
        console.error(
          `Error listing all chunks for source_id ${source_id} from vector db `,
          error.message
        );
      } else {
        console.error('Unexpected error', error);
      }
      throw error;
    }
  }

  async hardDeleteNameSpace(nameSpace: string) {
    try {
      await this.pc_index.namespace(nameSpace).deleteAll();
      console.log(`Deleted all records in namespace ${nameSpace}`);
    } catch (err) {
      throw err;
    }
  }
}

const transformAbsolutePath = async (
  path: string
): Promise<BaseNode<Metadata>[]> => {
  const documents = await loadDocuments([path]);
  // const parser = new SentenceSplitter();
  // const nodes = parser.getNodesFromDocuments(documents);
  const vectorStore = new PineconeVectorStore({
    indexName: 'sources',
    namespace: 'lama',
  });

  const pipeline = new IngestionPipeline({
    transformations: [
      new SentenceSplitter({ chunkSize: 512, chunkOverlap: 20 }),
      new TitleExtractor(),
      new OpenAIEmbedding(),
      // new OffsetExtractor(),
      // new KeywordExtractor(),
    ],
    // vectorStore,
  });

  const nodes = await pipeline.run({ documents: [documents[0]] });

  console.log(nodes.length, 'added to vector store');

  return nodes;

  // let currentOffset = 0;
  // nodes.forEach((node, index) => {
  //   const start = currentOffset;
  //   currentOffset += node.text.length;
  //   const end = currentOffset;
  //   node.metadata['start'] = start;
  //   node.metadata['end'] = end;
  // });
  //
  // console.log(nodes.length);
  // return nodes;
};

// const transformBuffer = async (
//   buffer: Buffer<ArrayBufferLike>
// ): Promise<BaseNode<Metadata>[]> => {
//   console.log('bufferSize', buffer.byteLength);
//
//   const document = new Document({
//     text: buffer.toString('utf-8'),
//     id_: 'buffer',
//   });
//
//   const pipeline = new IngestionPipeline({
//     transformations: [
//       new SentenceSplitter({ chunkSize: 1024, chunkOverlap: 20 }),
//       new OpenAIEmbedding(),
//     ],
//   });
//
//   return await pipeline.run({ documents: [document] });
// };

// const knowledgeIndex = new KnowledgeIndex('sources');
//
// knowledgeIndex.hardDeleteSource('936d92ec-b108-4f03-97b9-47332290a6e3');

// transformAbsolutePath('sam_altman_essay').then((nodes) => {
//   console.log(nodes);
// });

//TODO: create a PR for pineconeVB to geenrate vector id with prefix
export { KnowledgeIndex };

File: api/legacy/knowledge_index.py
# import os
# from typing import List, Optional
#
# from llama_index.core.vector_stores import MetadataFilters
# from llama_index.vector_stores.pinecone import PineconeVectorStore
# from pinecone import Pinecone
# from llama_index.core import (
#     VectorStoreIndex,
#     SimpleDirectoryReader,
# )
# from llama_index.core import StorageContext
#
#
# class KnowledgeIndex:
#     def __init__(self, index_name: str):
#         self.client = Pinecone(api_key=os.getenv("PINECONE_API_KEY", "your-api-key"))
#         self.pc_index = self.client.Index(index_name)
#
#     async def load_from_directories(
#         self, absolute_source_dir: str
#     ) -> Optional[VectorStoreIndex]:
#         file_list = await self.get_source_filenames(absolute_source_dir)
#         count = len(file_list)
#         print(f"Found {count} files")
#
#         try:
#             rdr = SimpleDirectoryReader(callback)
#             docs = await rdr.load_data(directory_path=absolute_source_dir)
#             pcvs = PineconeVectorStore(
#                 index_name="sources", namespace="lama", chunk_size=512
#             )
#             ctx = StorageContext.from_defaults(vector_store=pcvs)
#             return VectorStoreIndex.from_documents(docs, storage_context=ctx)
#         except Exception as err:
#             print(f"Error in load directories: {err}")
#             return None
#
#     async def upload_to_index(
#         self, source_meta: dict, buffer: bytes
#     ) -> Optional[VectorStoreIndex]:
#         source_id = source_meta["sourceId"]
#         source_name = source_meta["sourceName"]
#
#         try:
#             document = Document(
#                 text=buffer.decode("utf-8"),
#                 metadata={"source_id": source_id, "source_name": source_name},
#             )
#             pcvs = PineconeVectorStore(
#                 index_name="sources", namespace="buffers", chunk_size=1024
#             )
#             ctx = await StorageContext.from_defaults(vector_store=pcvs)
#             return await VectorStoreIndex.from_documents(
#                 [document], storage_context=ctx
#             )
#         except Exception as err:
#             print(err)
#             raise err
#
#     async def query(self, query: str):
#         pcvs = PineconeVectorStore(
#             index_name="sources", namespace="buffers", chunk_size=512
#         )
#         index = await VectorStoreIndex.from_vector_store(pcvs)
#         return index.as_query_engine().query(query=query)
#
#     async def query_documents(self, document_ids: List[str], query: str):
#         pcvs = PineconeVectorStore(
#             index_name="sources", namespace="buffers", chunk_size=512
#         )
#         index = await VectorStoreIndex.from_vector_store(pcvs)
#         filters = MetadataFilters(
#             filters=[{"key": "source_id", "value": document_ids, "operator": "in"}]
#         )
#         query_engine = index.as_query_engine(pre_filters=filters, similarity_top_k=5)
#         return query_engine.query(query=query)
#
#     async def summarize_source(self, source_id: str):
#         pcvs = PineconeVectorStore(
#             index_name="sources", namespace="buffers", chunk_size=512
#         )
#         index = await VectorStoreIndex.from_vector_store(pcvs)
#         filters = MetadataFilters(
#             filters=[{"key": "source_id", "value": [source_id], "operator": "in"}]
#         )
#         query_engine = index.as_query_engine(pre_filters=filters, similarity_top_k=10)
#         return query_engine.query(query="summarize the article")
#
#     async def hard_delete_source(self, ref_doc_id: str):
#         try:
#             pcvs = PineconeVectorStore(
#                 index_name="sources", namespace="buffers", chunk_size=512
#             )
#             index = await VectorStoreIndex.from_vector_store(pcvs)
#             print("index", index)
#             resp = await index.delete_ref_doc(ref_doc_id)
#             print(resp)
#         except Exception as error:
#             print(f"Error deleting doc_id {ref_doc_id} from vector db: {error}")
#             raise error
#
#     async def list_all_chunks_for_document(self, source_id: str):
#         try:
#             ns = self.pc_index.namespace("buffers")
#             res = await ns.list({"source_id": {"$eq": source_id}})
#             print(res)
#         except Exception as error:
#             print(
#                 f"Error listing all chunks for source_id {source_id} from vector db: {error}"
#             )
#             raise error
#
#     async def hard_delete_namespace(self, namespace: str):
#         try:
#             await self.pc_index.namespace(namespace).delete_all()
#             print(f"Deleted all records in namespace {namespace}")
#         except Exception as err:
#             raise err
#
#     async def get_source_filenames(self, source_dir: str) -> List[str]:
#         # Implement the logic to get source filenames
#         pass
#
#
# # async def transform_absolute_path(path: str) -> List[BaseNode[Metadata]]:
# #     documents = await load_documents([path])
# #     vector_store = PineconeVectorStore(index_name="sources", namespace="lama")
# #     pipeline = IngestionPipeline(
# #         transformations=[
# #             SentenceSplitter(chunk_size=512, chunk_overlap=20),
# #             TitleExtractor(),
# #             OpenAIEmbedding(),
# #         ]
# #     )
# #     nodes = await pipeline.run(documents=[documents[0]])
# #     print(f"{len(nodes)} added to vector store")
# #     return nodes
